src/corehost/cli/fxr/fx_muxer.cpp

```diff
@@ -702,102 +774,321 @@ int fx_muxer_t::execute(
 
 namespace
 {
-    int get_delegate_from_runtime(
-        const pal::string_t& impl_dll_dir,
-        corehost_init_t* init,
-        coreclr_delegate_type type,
-        void** delegate)
-    {
-        pal::dll_t corehost;
-        hostpolicy_contract host_contract{};
-        corehost_get_delegate_fn coreclr_delegate = nullptr;
-
-        int code = load_hostpolicy(impl_dll_dir, &corehost, host_contract, "corehost_get_coreclr_delegate", &coreclr_delegate);
-        if (code != StatusCode::Success)
-        {
-            trace::error(_X("This component must target .NET Core 3.0 or a higher version."));
-            return code;
-        }
-
-        {
-            propagate_error_writer_t propagate_error_writer_to_corehost(host_contract.set_error_writer);
-
-            const host_interface_t& intf = init->get_host_init_data();
-
-            if ((code = host_contract.load(&intf)) == StatusCode::Success)
-            {
-                code = coreclr_delegate(type, delegate);
-            }
-        }
-
-        return code;
-    }
-
     int get_init_info_for_component(
         const host_startup_info_t &host_info,
         host_mode_t mode,
         pal::string_t &runtime_config_path,
-        pal::string_t &impl_dir,
-        std::unique_ptr<corehost_init_t> &init)
+        /*out*/ pal::string_t &hostpolicy_dir,
+        /*out*/ std::unique_ptr<corehost_init_t> &init)
     {
         // Read config
         fx_definition_vector_t fx_definitions;
         auto app = new fx_definition_t();
         fx_definitions.push_back(std::unique_ptr<fx_definition_t>(app));
 
-        runtime_config_t::settings_t override_settings;
+        const runtime_config_t::settings_t override_settings;
         int rc = read_config(*app, host_info.app_path, runtime_config_path, override_settings);
         if (rc != StatusCode::Success)
             return rc;
 
-        runtime_config_t app_config = app->get_runtime_config();
+        const runtime_config_t app_config = app->get_runtime_config();
         bool is_framework_dependent = app_config.get_is_framework_dependent();
         if (is_framework_dependent)
```

Comment:
We should fail for self-contained components.
* SDK will never produce such a thing
* While in theory they might work currently the code actively blocks them in `hostpolicy` - https://github.com/dotnet/core-setup/blob/6dc96953da07af2e87254716491a0e37b421e7ea/src/corehost/cli/hostpolicy/deps_resolver.cpp#L504 This code means that for `libhost` we ignore the `.deps.json` for the component, but there are no frameworks in self-contained so the rest of the method will not load any other `.deps.json` either and we'll end up with empty TPA. I haven't tried this, but it will likely fail to find `coreclr` or some such really weird error.
* It's not a supported scenario because such components would only work if loaded first into a process without CoreCLR, we can't ever load them into a process which already has a CoreCLR.

We already block this for secondary contexts. And while from the point of view of hostfxr it is "valid" to create a primary context from self-contained, I think for now we should block this to avoid the confusion mentioned above.

I do realize that the code did behave like this before your change - but that was also wrong.